PublicMapping/districtbuilder
# Add rectangle selection tool

## Overview

Add rectangle selection tool

### Demo

![db_rectangle_selection](https://user-images.githubusercontent.com/2926237/87166141-241ebb00-c299-11ea-8b51-92264bf7fa96.gif)

### Notes

If you look back through the commit history you can see the whole saga of trying to use 3rd party plugins for this as suggested in the issue until I realized it could be done all within Mapbox and there was even an example in the docs shown us how. I think not adding any new deps is definitely the better way to go (and the plugins I was looking at don't do everything we wanted anyway).

This includes a change in code organization, adding a new `map/` directory within `components/` for all code that directly supports `Map.tsx`, including breaking out each selection tool into its own file (see 4682fab). This seems like a nice way to organize things but I'm not married to it if people have other ideas.

This ports over some styles for the buttons to make it approximate what's in the prototype. It's not perfect but it's pretty close which @jfrankl said is good enough for now.

The code for the rectangle selection tool is largely borrowed from the Mapbox docs example. The code is imperative and mutation-y, but so is Mapbox. I mostly left it as is, not wanting to spend the time to change the style of the code and in so doing potentially break stuff. Having the code be separated out in its own file makes everything a bit easier to digest and maintain (any "messiness" is cloistered off). It should also make it easy to add new selection tools going forward.

One other thing regarding the design of the selection tool objects that implement `ISelectionTool`: I considered using classes for this, because there is state that is shared between enabling and disabling (namely event handlers which needed to be added and removed correctly) and yet these selection tools are singletons so instantiating classes seemed to make less sense. Thus, with this design we simply set properties on `this` to share that state. I think the simplicity of what these do (`enable` does the thing, `disable` stops doing the thing and cleans up) should largely avoid state issues with this design. 

Regarding disabling lint errors, I wasn't able to disable them in a fine-grained way as I would have liked (see fff25ff)

## Testing Instructions
* Select geounits via default selection tool and rectangle tool, separately and in conjunction
* Switching back and forth between the two modes and ensure that state is being managed properly (no weird behavior around geounit selection)
* Test selection and deselection (rectangle tool should only select, not deselect, as per the prototype's behavior)
* Check that when in default selection mode, the cursor is a grabby hand for panning when not hovered over the features and then becomes a pointer after hovering
* Check that when in rectangle selection mode, the cursor is always the crosshairs (can't pan in this mode)

Closes #110
